pymuonsuite/muairss.py

```diff
@@ -234,43 +127,33 @@ def save_muairss_collection(struct, params, batch_path=''):
     dc.info['muairss_params'] = dict(params)
 
     # Output folder
-    out_path = safe_create_folder(os.path.join(batch_path, params['out_folder']))
+    out_path = safe_create_folder(os.path.join(batch_path,
+                                  params['out_folder']))
 
     if not out_path:
         raise RuntimeError('Could not create folder {0}')
 
-    # Now save in the appropriate format
-    save_formats = {
-        'castep': castep_write_input,
-        'dftb+': dftb_write_input,
-        'uep': uep_write_input
+    io_formats = {
+        'castep': ReadWriteCastep(params),
+        'dftb+': ReadWriteDFTB(params),
+        'uep': ReadWriteUEP(params)
     }
 
-    # Which calculators?
     calcs = [s.strip().lower() for s in params['calculator'].split(',')]
     if 'all' in calcs:
-        calcs = save_formats.keys()
-
-    # Make the actual calculators
-    make_calcs = {
-        'castep': create_muairss_castep_calculator,
-        'dftb+': create_muairss_dftb_calculator,
-        'uep': create_muairss_uep_calculator
-    }
-
-    calcs = {c: make_calcs[c](struct, params=params, calc=struct.calc)
-             for c in calcs}
+        calcs = io_formats.keys()
 
     # Save LICENSE file for DFTB+ parameters
     if 'dftb+' in calcs:
         from pymuonsuite.data.dftb_pars import get_license
         with open(os.path.join(out_path, 'dftb.LICENSE'), 'w') as f:
             f.write(get_license())
 
-    for cname, calc in calcs.items():
+    for cname in calcs:
         calc_path = os.path.join(out_path, cname)
-        dc.save_tree(calc_path, save_formats[cname], name_root=params['name'],
-                     opt_args={'calc': calc, 'script': params['script_file']},
+        dc.save_tree(calc_path, io_formats[cname].write,
```

Comment:
Small detail, but this way we're creating three `ReadWrite` objects even if we won't necessarily need them all. You could as well just have only the classes in `io_formats`, and then here:

```python
rw = io_formats[cname](params)
dc.save_tree(calc_path, rw.write, ...)
```

Reply:
Resolved here: https://github.com/muon-spectroscopy-computational-project/pymuon-suite/pull/7/commits/a1306f6c12edd108293ada3bc090abda739a0ac6